src/replication.c

```diff
@@ -850,112 +322,83 @@ update_replica_entry(spec_t *spec, replica_t *replica, int iofd)
 	replica->pool_guid = ack_data->pool_guid;
 	replica->zvol_guid = ack_data->zvol_guid;
 
-	replica->least_recvd = 0;
-	replica->wrio_seq = 0;
-	replica->rrio_seq = 0;
 	replica->spec = spec;
 	replica->io_resp_hdr = (zvol_io_hdr_t *)malloc(sizeof(zvol_io_hdr_t));
-	replica->io_resp_data = NULL;
 	replica->io_state = READ_IO_RESP_HDR;
 	replica->io_read = 0;
-	replica->removed = false;
-
-	MTX_LOCK(&spec->rq_mtx);
-
-	replica->id = spec->replica_count;
-	spec->replica_count++;
-	TAILQ_REMOVE(&spec->rwaitq, replica, r_waitnext);
-	spec->degraded_rcount++;
-	TAILQ_INSERT_TAIL(&spec->rq, replica, r_next);
-	if(spec->replica_count == 1)
-		pthread_cond_signal(&spec->rq_cond);
-	update_volstate(spec);
-
-	MTX_UNLOCK(&spec->rq_mtx);
 
 	rio = (zvol_io_hdr_t *)malloc(sizeof(zvol_io_hdr_t));
 	rio->opcode = ZVOL_OPCODE_HANDSHAKE;
 	rio->io_seq = 0;
 	rio->offset = 0;
-	rio->len    = strlen(spec->lu->volname);
+	rio->len = strlen(spec->lu->volname);
 	rio->version = REPLICA_VERSION;
-	write(replica->iofd, rio, sizeof(zvol_io_hdr_t));
-	write(replica->iofd, spec->lu->volname, strlen(spec->lu->volname));
-	free(rio);
 
-	event.data.fd = replica->iofd;
-	event.events = EPOLLIN | EPOLLET;
-	rc = epoll_ctl(spec->receiver_epfd, EPOLL_CTL_ADD, iofd, &event);
-	if (rc == -1) {
-		/*
-		 * TODO: need to cleanup replica and also update volstate here
-		 */
-		return NULL;
+	if (write(replica->iofd, rio, sizeof(zvol_io_hdr_t)) != sizeof(zvol_io_hdr_t)) {
+		REPLICA_ERRLOG("failed to send io hdr to replica\n");
+		goto replica_error;
 	}
 
-	replica->sender_epfd = epoll_create1(0);
-	event.data.fd = replica->iofd;
-	event.events = EPOLLOUT;
-	rc = epoll_ctl(replica->sender_epfd, EPOLL_CTL_ADD, replica->iofd, &event);
-	if (rc == -1) {
-		REPLICA_LOG("epoll_ctl_add failed errno:%d\n", errno);
-		/*
-		 * TODO: need to cleanup replica and also update volstate here
-		 */
-		return NULL;
+	if (write(replica->iofd, spec->lu->volname,
+		strlen(spec->lu->volname)) != (ssize_t)strlen(spec->lu->volname)) {
+		REPLICA_ERRLOG("failed to send volname to replica\n");
+		goto replica_error;
 	}
 
-	rc = pthread_create(&replica_sender_thread, NULL, &replica_sender,
-			(void *)replica);
-	if (rc != 0) {
-		ISTGT_ERRLOG("pthread_create(replicator_thread) failed\n");
-		/*
-		 * TODO: need to cleanup replica and also update volstate here
-		 */
-		return NULL;
+	if (read(replica->iofd, rio, sizeof (*rio)) != sizeof (*rio)) {
+		REPLICA_ERRLOG("failed to read handshake response from replica\n");
+		goto replica_error;
 	}
 
-	return replica;
-}
+	if(init_mempool(&replica->cmdq, rcmd_mempool_count, sizeof (rcmd_t), 0,
```

Comment:
It doesn't seem to be having MP, SC setting for this mempool.
Please verify the flags for other mempools also.